### DISCUSSION TITLE
Using external packages when using hooks

### DISCUSSION THREAD

**Discussion by txstc55** [pocketbase/pocketbase#3332] [Q&A] [answered]
Hi, if I have an external node package I want to call before an update using the hook, say something like getting the current weather, or simply, I want to send an https request, is it possible with the current pocket base?

I'm asking this question because I want to completely strip away a dedicated backend, and implement everything in just hooks. But I'm not sure if adding external node packages is possible.

**Comment by ganigeorgiev** [comment, marked as answer]
_I assume you mean the JS `pb_hooks`._

External node modules may not work because the embedded PocketBase JS engine is not a node/browser environment and a lot of runtime specific apis like `fs`, `fetch`, etc. are not available.

If you haven't already, you can also check the [Caveats and limitations section](https://pocketbase.io/docs/js-overview/#caveats-and-limitations) (and more specifically [Loading modules](https://pocketbase.io/docs/js-overview/#loading-modules)).

> I want to send an https request, is it possible with the current pocket base?

See https://pocketbase.io/docs/js-sending-http-requests/.

**Reply by txstc55** [reply]
Thank you! So if I'm using a go backend this is not a limitation?

**Reply by ganigeorgiev** [reply]
Yes, when using Go you have more control over the application flow and you should be able to integrate with any external Go package (_PocketBase itself is distributed as regular Go package and it lives as part of your Go program; see the example in https://pocketbase.io/docs/go-overview/_).

**Reply by benallfree** [reply]
@ganigeorgiev This conversation started over at https://github.com/benallfree/pockethost/discussions/288.

I think a [fetch](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) wrapper for `$http.send` would be helpful. It would also be nice, from [pockethost.io](https://pockethost.io)'s standpoint, to be able to whitelist which domains it is allowed to contact.

**Reply by ganigeorgiev** [reply]
@benallfree 
Implementing the entire `fetch` spec (and its related apis like Request, Response, streams, etc.) seems very involving to me and I'm not sure if it is worth the dev effort at the moment.

Also note that this is not something that need to be implemented specifically for PocketBase and if someone has the time and wants to pick it they can contribute directly to [goja](https://github.com/dop251/goja) / [goja_nodejs](https://github.com/dop251/goja_nodejs) (_but I'd advice first to discuss it with the maintainer_).

> It would also be nice, from [pockethost.io](https://pockethost.io/)'s standpoint, to be able to whitelist which domains it is allowed to contact.

I fully understand your use case but sandboxing for the JS `pb_hooks` for now is not planned as it will be very complicated, error prone and for most users who selfhost PocketBase this will not have any benefit and may even cause confusion depending how the sandbox configuration is implemented.

Also keep in mind that users are allowed to execute shell commands using the [`$os` binds](https://pocketbase.io/jsvm/modules/_os.html), so even if we restrict `$http.send`, technically they still could perform HTTP requests using `curl`, `wget` or anything else available on the OS to the user that started the process.
If you are looking for a way to sandbox your users my recommendation would be to research approaches that works independently from PocketBase, for example somehow define via ufw or similar rules that will allow/disallow certain outbound connections and rate limits for the individual user container.
